core/src/transport/circuit_breaking.rs

```diff
@@ -0,0 +1,458 @@
+        let circuit_addr = CircuitAddr::from(addr.clone());
+        let removed = self.addresses.lock().remove(&circuit_addr);
+        debug!("Closed circuit {:?}", removed);
+    }
+
+    fn close_expired_circuits(&self) {
+        let now = Instant::now();
+        let mut addresses = self.addresses.lock();
+        addresses.retain(|_, a| a.expiration > now);
+        debug!("Closed expired circuits")
+    }
+}
+
+/// Wraps a `Multiaddr` whose `PeerId` (if any) has been
+/// [decapsulated](https://github.com/libp2p/specs/blob/master/addressing/README.md#decapsulation).
+///
+/// In large public networks (e.g. IPFS) many nodes advertise LAN local
+/// addresses. In most cases nodes don't share the same LAN and thus dialing
+/// these addresses fails.
+///
+/// Ignoring the `PeerId` in addresses prevents dialing e.g. `192.168.2.108`
+/// multiple times in a row.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+struct CircuitAddr(Multiaddr);
+
+impl From<Multiaddr> for CircuitAddr {
+    fn from(mut addr: Multiaddr) -> Self {
+        // Check if `addr` contains `Protocol::P2p(_)`.
+        let contains_p2p = addr
+            .iter()
+            .any(|protocol| matches!(protocol, Protocol::P2p(_)));
+        if !contains_p2p {
+            return Self(addr);
+        }
+
+        // Decapsulate `Protocol::P2p(_)`.
+        loop {
+            match addr.pop() {
+                Some(Protocol::P2p(_)) | None => break,
+                _ => {}
+            }
+        }
+        Self(addr)
+    }
+}
+
+impl<TInner> CircuitBreaking<TInner> {
+    pub fn new(inner: TInner) -> Self {
+        CircuitBreaking {
+            inner,
+            state: Arc::new(CircuitBreakerState::new()),
+        }
+    }
+}
+
+impl<TInner> Transport for CircuitBreaking<TInner>
+where
+    TInner: Transport,
+    TInner::Error: 'static,
+{
+    type Output = TInner::Output;
+    type Error = TInner::Error;
+    type Listener = TInner::Listener;
+    type ListenerUpgrade = TInner::ListenerUpgrade;
+    type Dial = CircuitBreakingDial<TInner>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        self.inner.listen_on(addr)
+    }
+
+    /// Dials immediately if there is no open circuit related to the
+    /// `CircuitAddr` corresponding to `addr`. Otherwise it waits before
+    /// dialing, see [`CircuitBreaking`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if `addr` becomes unsupported while waiting for its circuit to
+    /// close. For an open circuit to exist, `addr` must have been supported
+    /// when it was dialed initially.
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        // Purge expired addresses.
+        self.state.close_expired_circuits();
+
+        let inner = match self.state.get_wait_duration(&addr) {
+            None => {
+                debug!("Dialing address immediately (no open circuit): {:?}", addr);
```

Comment:
Multiaddresses support formatting with `Display`!
```suggestion
                debug!("Dialing address immediately (no open circuit): {}", addr);
```

Also, I think rewording this might be a bit clearer:

```suggestion
                debug!("No open circuit for address {}, dialing immediately", addr);
```